File: Java/992.java
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
    public int subarraysWithKDistinct(int[] A, int K) {
        int n=A.length;
        int l=0;
        int[] count=new int[n+1];
        int unique=0,bias=0,res=0;
        for(int r=0;r<n;r++){
            if(++count[A[r]]==1)unique++;
            if(unique>K){
                bias=0;
                --count[A[l++]];
                unique--;
            }
            while(count[A[l]]>1){
                --count[A[l++]];
                bias++;
             
            }
            if(unique==K)res+=bias+1;
        }
        return res;
        
    }
}
__________________________________________________________________________________________________
sample 39464 kb submission
class Solution {
    public int subarraysWithKDistinct(int[] A, int K) {
        if (A == null || A.length < K) {
            return 0;
        }
        
        int[] hash = new int[A.length + 1];
        
        int l = 0, results = 0, count = 0, result = 1;
        for (int r = 0; r < A.length; ++r) {
            hash[A[r]]++;
            
            if (hash[A[r]] == 1) {
                count++;
            }
            
            if (count > K) {
                count--;
                result = 1;
                hash[A[l]]--;
                l++;
            }
            
            while (hash[A[l]] > 1) {
                result++;
                hash[A[l]]--;
                l++;
            }
            
            if (count == K) {
                results += result;
            }
        }
        
        return results;
    }
}
__________________________________________________________________________________________________

File: Python3/992.py
__________________________________________________________________________________________________
sample 396 ms submission
import copy
class Solution:
    def subarraysWithKDistinct(self, A: List[int], K: int) -> int:        
        if not A:
            return 0
        if K > len(set(A)):
            return 0
        
        dic = {A[0]: 0}
        i = 0
        j = 0
        res = 0
        
        while len(dic) < K:
            j += 1
            dic[A[j]] = j
        
        mk, mv = min(dic.items(), key = lambda x: x[1])
        
        while j < len(A):
            dic[A[j]] = j
            l = len(dic)
            # print(j, dic, mk, mv, A[mv])
            # print(dic[A[mv]] )
            
            if A[j] == mk:
                while dic[mk] != mv:
                    mv += 1
                    mk = A[mv]
                
            elif l > K:
                del dic[mk]
                mv += 1
                i = mv
                mk = A[mv]
                while dic[mk] != mv:
                    mv += 1
                    mk = A[mv]
                    
            res += mv - i + 1
                
            j += 1
            # print(i, j, l, dic, mi, mv, res)
                
        return res
            
__________________________________________________________________________________________________

__________________________________________________________________________________________________
